From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ritu Chaudhary <rituc@nvidia.com>
Date: Thu, 17 Apr 2025 10:45:14 +0000
Subject: [PATCH 217/254] NVIDIA: SAUCE: ASoC: tegra: Add Mixer Fade controls

Fade controls set the gain and N3 (total duration in samples) for the
specified mixer streams and provide status in terms of whether fade is
inactive(-1), active(0) or complete(1). Once status is read and fade is
complete, the status will be set to inactive for that stream.

N3 represents the total duration parameter used in gain curve calculations
of the form n/N3 or (N3-n)/N3, where n is the current sample position.

Usage Example:
  amixer -c <card> cset name="MIXER1 Fade" "1,12000,1024,2,15000,2048"
    Fade control accepts total 10(streams)x3(rx id, gain, duration N3)
    values.

  amixer -c <card> cget name="MIXER1 Fade Status" gives below output
    "0,1,-1,-1,-1,-1,-1,-1,-1,-1"

Bug 4870208

Change-Id: I37dfc529221143522dff23fd422c1d2f1a226a7a
Signed-off-by: Sheetal <sheetal@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3342845
Reviewed-by: Mohan kumar <mkumard@nvidia.com>
Reviewed-by: svcacv <svcacv@nvidia.com>
Reviewed-by: Sameer Pujar <spujar@nvidia.com>
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
(cherry picked from commit b140f55a0d5c062a75b8230d81a374987ad70f18)

diff --git a/sound/soc/tegra/tegra210_mixer.c b/sound/soc/tegra/tegra210_mixer.c
index 69598271ab5e4..72e2bd21f5275 100644
--- a/sound/soc/tegra/tegra210_mixer.c
+++ b/sound/soc/tegra/tegra210_mixer.c
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: GPL-2.0-only
-// SPDX-FileCopyrightText: Copyright (c) 2021-2024 NVIDIA CORPORATION & AFFILIATES.
+// SPDX-FileCopyrightText: Copyright (c) 2021-2025 NVIDIA CORPORATION & AFFILIATES.
 // All rights reserved.
 //
 // tegra210_mixer.c - Tegra210 MIXER driver
@@ -151,10 +151,17 @@ static int tegra210_mixer_configure_gain(struct snd_soc_component *cmpnt,
 	for (i = 0; i < NUM_DURATION_PARMS; i++) {
 		int val;
 
-		if (instant_gain)
+		if (instant_gain) {
 			val = 1;
-		else
-			val = gain_params.duration[i];
+		} else {
+			if (i == DURATION_N3_ID)
+				val = mixer->duration[id];
+			else if (i == DURATION_INV_N3_ID)
+				val = (BIT(31) * BIT(TEGRA210_MIXER_PRESCALAR)) /
+					mixer->duration[id];
+			else
+				val = gain_params.duration[i];
+		}
 
 		err = tegra210_mixer_write_ram(mixer,
 					       REG_DURATION_PARAM(reg, i),
@@ -173,6 +180,123 @@ static int tegra210_mixer_configure_gain(struct snd_soc_component *cmpnt,
 	return err;
 }
 
+static int tegra210_mixer_put_fade(struct snd_kcontrol *kcontrol,
+				   struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *cmpnt = snd_kcontrol_chip(kcontrol);
+	struct tegra210_mixer *mixer = snd_soc_component_get_drvdata(cmpnt);
+	u32 rx_id, rx_gain, rx_duration;
+	int id, base, err = 0;
+
+	/* Process array of RX stream parameters (id, gain, duration) */
+	for (id = 0; id < TEGRA210_MIXER_RX_MAX; id++) {
+		base = 3 * id;
+		rx_id = ucontrol->value.integer.value[base];
+		rx_gain = ucontrol->value.integer.value[base + 1];
+		rx_duration = ucontrol->value.integer.value[base + 2];
+
+		/* Array ends here, user doesn't require to pass all RX id configuration */
+		if (rx_id == 0 && rx_gain == 0 && rx_duration == 0)
+			break;
+
+		if (rx_id < 1 || rx_id > TEGRA210_MIXER_RX_MAX) {
+			dev_err(cmpnt->dev, "Invalid mixer rx index %d\n", rx_id);
+			err = -EINVAL;
+			goto exit;
+		}
+
+		/* Convert 1-based rx_id to 0-based array index */
+		rx_id = rx_id - 1;
+
+		/* Only update if values are different */
+		if (mixer->duration[rx_id] != rx_duration ||
+		    mixer->gain_value[rx_id] != rx_gain) {
+			err = tegra210_mixer_configure_gain(cmpnt, rx_id, false);
+			if (err) {
+				dev_err(cmpnt->dev,
+					"Failed to configure fade for channel %d\n", rx_id);
+				goto exit;
+			}
+			err = 1;
+			mixer->duration[rx_id] = rx_duration;
+			mixer->gain_value[rx_id] = rx_gain;
+		}
+
+		if (!mixer->in_fade[rx_id]) {
+			pm_runtime_get_sync(cmpnt->dev);
+			err = regmap_update_bits(mixer->regmap,
+						 MIXER_REG(TEGRA210_MIXER_RX1_CTRL, rx_id),
+						 TEGRA210_MIXER_SAMPLE_COUNT_ENABLE,
+						 TEGRA210_MIXER_SAMPLE_COUNT_ENABLE);
+			pm_runtime_put(cmpnt->dev);
+			if (err) {
+				dev_err(cmpnt->dev,
+					"Failed to enable sample count for channel %d\n", rx_id);
+				goto exit;
+			}
+			err = 1;
+			mixer->in_fade[rx_id] = true;
+		}
+	}
+
+exit:
+	return err;
+}
+
+static int tegra210_mixer_get_fade(struct snd_kcontrol *kcontrol,
+				   struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *cmpnt = snd_kcontrol_chip(kcontrol);
+	struct tegra210_mixer *mixer = snd_soc_component_get_drvdata(cmpnt);
+	u32 id, base, index = 0;
+
+	for (id = 0; id < TEGRA210_MIXER_RX_MAX; id++) {
+		if (mixer->in_fade[id]) {
+			base = 3 * index;
+			ucontrol->value.integer.value[base] = id + 1;
+			ucontrol->value.integer.value[base + 1] = mixer->gain_value[id];
+			ucontrol->value.integer.value[base + 2] = mixer->duration[id];
+			index++;
+		}
+	}
+
+	return 0;
+}
+
+static int tegra210_mixer_get_fade_status(struct snd_kcontrol *kcontrol,
+					  struct snd_ctl_elem_value *ucontrol)
+{
+	struct snd_soc_component *cmpnt = snd_kcontrol_chip(kcontrol);
+	struct tegra210_mixer *mixer = snd_soc_component_get_drvdata(cmpnt);
+	u32 count;
+	int id;
+
+	pm_runtime_get_sync(cmpnt->dev);
+
+	for (id = 0; id < TEGRA210_MIXER_RX_MAX; id++) {
+		if (mixer->in_fade[id]) {
+			regmap_read(mixer->regmap, MIXER_REG(TEGRA210_MIXER_RX1_SAMPLE_COUNT, id),
+				    &count);
+
+			if (count >= mixer->duration[id]) {
+				ucontrol->value.integer.value[id] = TEGRA210_MIXER_FADE_COMPLETE;
+				regmap_update_bits(mixer->regmap,
+						   MIXER_REG(TEGRA210_MIXER_RX1_CTRL, id),
+						   TEGRA210_MIXER_SAMPLE_COUNT_ENABLE, 0);
+				mixer->in_fade[id] = false;
+			} else {
+				ucontrol->value.integer.value[id] = TEGRA210_MIXER_FADE_ACTIVE;
+			}
+		} else {
+			ucontrol->value.integer.value[id] = TEGRA210_MIXER_FADE_IDLE;
+		}
+	}
+
+	pm_runtime_put(cmpnt->dev);
+
+	return 0;
+}
+
 static int tegra210_mixer_get_gain(struct snd_kcontrol *kcontrol,
 				   struct snd_ctl_elem_value *ucontrol)
 {
@@ -396,6 +520,46 @@ ADDER_CTRL_DECL(adder3, TEGRA210_MIXER_TX3_ADDER_CONFIG);
 ADDER_CTRL_DECL(adder4, TEGRA210_MIXER_TX4_ADDER_CONFIG);
 ADDER_CTRL_DECL(adder5, TEGRA210_MIXER_TX5_ADDER_CONFIG);
 
+static int tegra210_mixer_param_info(struct snd_kcontrol *kcontrol,
+				     struct snd_ctl_elem_info *uinfo)
+{
+	struct soc_bytes *params = (void *)kcontrol->private_value;
+
+	/* Set default parameters for integer type controls */
+	if (params->mask == SNDRV_CTL_ELEM_TYPE_INTEGER) {
+		uinfo->value.integer.min = 0;
+		uinfo->value.integer.max = U32_MAX;
+	}
+
+	/* Special case for status controls */
+	if (strstr(kcontrol->id.name, "Status"))
+		params->num_regs = 10;
+	else
+		params->num_regs = 30;
+
+	/* Set control info */
+	uinfo->type = params->mask;
+	uinfo->count = params->num_regs;
+
+	return 0;
+}
+
+#define MIXER_FADE_CTRL(xname, xbase, xnum_regs, xget, xput, xaccess) \
+{ \
+	.iface = SNDRV_CTL_ELEM_IFACE_MIXER, \
+	.info =  tegra210_mixer_param_info, \
+	.name = xname, \
+	.access = xaccess, \
+	.get = xget, \
+	.put = xput, \
+	.private_value = (unsigned long)&(struct soc_bytes)	\
+	{ \
+		.base = xbase, \
+		.num_regs = xnum_regs, \
+		.mask = xaccess \
+	} \
+},
+
 #define GAIN_CTRL(id)	\
 	SOC_SINGLE_EXT("RX" #id " Gain Volume",			\
 		       MIXER_GAIN_CFG_RAM_ADDR((id) - 1), 0,	\
@@ -418,6 +582,11 @@ static const struct snd_kcontrol_new tegra210_mixer_gain_ctls[] = {
 	GAIN_CTRL(8)
 	GAIN_CTRL(9)
 	GAIN_CTRL(10)
+
+	MIXER_FADE_CTRL("Fade", 0x0, 30, tegra210_mixer_get_fade, tegra210_mixer_put_fade,
+			SNDRV_CTL_ELEM_ACCESS_READWRITE)
+	MIXER_FADE_CTRL("Fade Status", 0x0, 10, tegra210_mixer_get_fade_status, NULL,
+			SNDRV_CTL_ELEM_ACCESS_READ)
 };
 
 static const struct snd_soc_dapm_widget tegra210_mixer_widgets[] = {
@@ -652,6 +821,7 @@ static bool tegra210_mixer_volatile_reg(struct device *dev,
 	case TEGRA210_MIXER_GAIN_CFG_RAM_DATA:
 	case TEGRA210_MIXER_PEAKM_RAM_CTRL:
 	case TEGRA210_MIXER_PEAKM_RAM_DATA:
+	case TEGRA210_MIXER_RX1_SAMPLE_COUNT:
 		return true;
 	default:
 		return false;
@@ -705,8 +875,10 @@ static int tegra210_mixer_platform_probe(struct platform_device *pdev)
 	dev_set_drvdata(dev, mixer);
 
 	/* Use default gain value for all MIXER inputs */
-	for (i = 0; i < TEGRA210_MIXER_RX_MAX; i++)
+	for (i = 0; i < TEGRA210_MIXER_RX_MAX; i++) {
 		mixer->gain_value[i] = gain_params.gain_value;
+		mixer->duration[i] = gain_params.duration[DURATION_N3_ID];
+	}
 
 	regs = devm_platform_ioremap_resource(pdev, 0);
 	if (IS_ERR(regs))
diff --git a/sound/soc/tegra/tegra210_mixer.h b/sound/soc/tegra/tegra210_mixer.h
index a330530fbc613..030a6a0dab527 100644
--- a/sound/soc/tegra/tegra210_mixer.h
+++ b/sound/soc/tegra/tegra210_mixer.h
@@ -1,8 +1,8 @@
-/* SPDX-License-Identifier: GPL-2.0-only */
-/*
- * tegra210_mixer.h - Definitions for Tegra210 MIXER driver
+/* SPDX-License-Identifier: GPL-2.0-only
+ * SPDX-FileCopyrightText: Copyright (c) 2021-2025 NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
  *
- * Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+ * tegra210_mixer.h - Definitions for Tegra210 MIXER driver
  *
  */
 
@@ -79,12 +79,26 @@
 #define TEGRA210_MIXER_RX_LIMIT		(TEGRA210_MIXER_RX_MAX * TEGRA210_MIXER_REG_STRIDE)
 #define TEGRA210_MIXER_TX_MAX		5
 #define TEGRA210_MIXER_TX_LIMIT		(TEGRA210_MIXER_RX_LIMIT + (TEGRA210_MIXER_TX_MAX * TEGRA210_MIXER_REG_STRIDE))
+#define TEGRA210_MIXER_SAMPLE_COUNT_SHIFT	24
+#define TEGRA210_MIXER_SAMPLE_COUNT_ENABLE	BIT(TEGRA210_MIXER_SAMPLE_COUNT_SHIFT)
 
 #define REG_CFG_DONE_TRIGGER	0xf
 #define VAL_CFG_DONE_TRIGGER	0x1
 
 #define NUM_GAIN_POLY_COEFFS 9
-#define NUM_DURATION_PARMS 4
+
+#define TEGRA210_MIXER_PRESCALAR	6
+#define TEGRA210_MIXER_FADE_IDLE	(-1)
+#define TEGRA210_MIXER_FADE_ACTIVE	0
+#define TEGRA210_MIXER_FADE_COMPLETE	1
+
+enum {
+	DURATION_N1_ID,
+	DURATION_N2_ID,
+	DURATION_N3_ID,
+	DURATION_INV_N3_ID,
+	NUM_DURATION_PARMS,
+};
 
 struct tegra210_mixer_gain_params {
 	int poly_coeff[NUM_GAIN_POLY_COEFFS];
@@ -94,6 +108,8 @@ struct tegra210_mixer_gain_params {
 
 struct tegra210_mixer {
 	int gain_value[TEGRA210_MIXER_RX_MAX];
+	u32 duration[TEGRA210_MIXER_RX_MAX];
+	bool in_fade[TEGRA210_MIXER_RX_MAX];
 	struct regmap *regmap;
 };
 
-- 
2.43.0

